### ISSUE TITLE
RFC: Refactor consent flow

### ISSUE THREAD

**Issue by arekkas** [ory/hydra#578]
The consent flow is a center piece of how we authenticate users. It works by creating a trust relationship between hydra and the consent app (usually a login user interface). One issue with the consent flow is that setting up the client correctly and getting JWK + JWTs to work is quite a hassle even if an SDK exists. Additionally, the SDK can not be code generated because it relies on external libraries (like JWK and JWT libraries).

Here are two proposals that could alleviate this pain.

# Replacing the JWTs with a REST request

The first possibility is to replace the validation and singing of the consent challenge and response with REST requests. There would be one request to query information, and one request to verify authentication. The flow would initiate similarly to today:

1. A user agent makes an OAuth2 request to Hydra
2. Hydra redirects to the consent app, but instead of appending a consent challenge we append a unique, unguessable session id: `https://login.myapp.com/consent?session=<session-id>`
3. The consent app makes a REST request to query information
```
GET /oauth2/authorize/sessions/<session-id>
Authorization: bearer <priviledged-token>
```
```
{
  "audience": "foo",
  "scopes": "bar"
  // ...
}
```
4. The consent app authenticates the user and (optionally) asks for his/her consent.
5. The consent app makes another REST request to verify the access request:
```
POST /oauth2/authorize/sessions/<session-id>
Authorization: bearer <priviledged-token>

{
  "action": "allow", // could also be deny
  "scopes": // ... and so on
}
```
```
{
  "audience": "foo",
  "scopes": "bar"
  // ...
}
```
6. The consent app redirects the user back to Hydra: `https://hydra.myapp.com/oauth2/authorize?session=<sessino-id>`
7. Hydra extracts the information and uses that to forge an access token, and so on.

## Benefits

* We get rid of JWK and JWT complexity
* We get rid of an HTTP call to the JWK endpoint
* It is likely that we can code generate this (meaning an SDK for every language)
* We do not rely on RSA which has decreasing security with increasing computational power (or someone solving P=NP)
* We throw away a specification which is known to contain bugs in library implementations due to a vague spec.

## Downsides

* We still need to perform a client_credentials grant in order to query and post the session endpoint
* We can no longer hide the private key for the consent response from hydra (hydra only needs the public key to validate it)
* We have more network round trips, see [this comment](https://github.com/ory/hydra/issues/578#issuecomment-328882313)

### Neutral

* Trust relationship remains equal because the current way has also a REST-ful trust relationship (for fetching the JWKs)

### Ideas

To make set up of the consent app easier, we could add a command to the CLI that automates this task: `hydra clients create --template consent-app my-app-id`.

# Keep JWT dance but replace PKI with shared secret

Alternatively we could still rely on the consent challenge & response pattern but throw away the need for a public & private key pair and instead use a shared secret. This would allow the consent app to work without ever querying hydra for data and thus remove the pain of setting up a consent app client and doing the client_credentials grant for fetching JWKs.

## Upsides

* JWK dance / complexity is gone
* No more setting up of the consent app client because all we need is the shared secret

## Downsides

* We keep relying on JWTs and probably need to write SDKs by hand (?)
* Having only a shared key as security measure could promote insecure behaviour

## Neutral

* Shared secrets can be promoted as such, meaning that people won't accidentally include them on SPAs or mobile apps.

**Comment by arekkas** [comment]
Please comment on this issue and let me know what you think of the proposed ideas. I would also like to hear how the current consent flow has impacted your development (e.g. no problem - or - a lot of work)

**Comment by matteosuppo** [comment]
I had some difficulties implementing the consent flow, and now that it's working I'm not eager on doing it again, but it seems to me that this new consent flow could be easier to implement.

One downside is that it requires more network round-trips (and probably more trips to the database)

Old method:

get hydra 
get consent
post consent
get hydra

new method

get hydra
get consent
get session
post consent
post session
get hydra

**Comment by arekkas** [comment]
Yes that is true, but also only if the JWK is fetched only once and then cached for re-use. In a naive implementation, the consent app might call the JWK endpoint every time it requires the key. However, I added it to the list above because it's a good point you're making!

**Comment by waynerobinson** [comment]
Sorry it's taken me so long to look at this.

Even though we have a current working implementation, we occasionally run into some unusual edge-cases related to them.

However, we actually rely on some info in the challenge token for passing information to the consent app from our load-balancer, namely we need to extract the domain from the `redir` portion to know which partner the authentication is for. 

As long as there is some way to continue to pass information to consent app (say as part of the REST request to query that info), I'm for the change as it has the potential to simplify some of the hoops we have to jump through (although some of our use-cases are quite unique) to keep the challenge token around.

**Comment by arekkas** [comment]
@waynerobinson the direction it's currently headed (API) is more suitable to your use case than manually parsing the consent token. The API response will include important information which had to be kept out of the JWT to save space!

**Comment by janekolszak** [comment]
## Option 1: Replacing the JWTs with a REST request
### Speed:
- Private traffic between server nodes is generally free and fast. 
- It may be even better if Hydra is on the same physical machine. 
- We may use http instead of https. Possibly another RPC some day. Maybe Unix sockets?
- More data to send via HTTP and it won't be packed in a JWT
- More data send to the database
- Less time spent in signing / verifying signature

Nevertheless I guess it'll slow request handling down.

### Complexity
- We already have libraries for handling the JWT JWK magic. Rewriting is harder than leaving it be :)

### Security
- Opens up new attack pane - session id's introduce state that's kept in Hydra. 

The problem with micro services is that people tend to treat them like daemons in OS, but they're not. The RPC and availability is a nightmare.

## Option 2: Keep JWT dance but replace PKI with shared secret
- Probably easier if someone writes a new IdP

## Option 3: Use the database 
- Option 1 but without http.
- You already control the SDK, no complication for the user
- Access guarded by privileges on the database
- Works only if IdP has access to the Hydra's database. 
- Fast but doesn't feel right

## Option 4: Leave it be
- I'm all for this option since I can't rewrite my IdP :smile:

**Comment by arekkas** [comment]
Thank you for the sum up! A couple of notes from my side:

> More data to send via HTTP and it won't be packed in a JWT

JWT requires more bytes than regular REST traffic because we have signatures. This is the reason why the JWT claims are all short `aud`, `id_ext`, ... as we hit the URL length limit at some point.

> We already have libraries for handling the JWT JWK magic. Rewriting is harder than leaving it be :)

Without JWT/JWK I think it would be possible to auto.generate client libraries for all languages! Getting JWK to work is always a lot of work because usually JWT libraries use PEM encoded files.

> Nevertheless I guess it'll slow request handling down.

Maybe a bit, but not significantly.

> Option 3: Use the database
> Fast but doesn't feel right

Definitely.

> Option 4: Leave it be
> I'm all for this option since I can't rewrite my IdP 

Why not? Not a lot will change and I will supply an SDK for it :)

**Comment by janekolszak** [comment]
What will be in the SDK?

**Comment by ckong1991** [comment]
I'm more in favor of option 2.

For me, most of the time was spent learning about hydra's api and how the consent flow worked (e.g. which endpoints I needed to spin up, and how the data passed between hydra and the consent app) - writing my own functions didn't take too long, so although an SDK in option 1 would be useful, its benefits might not be game changing - the network roundtrips is a definite downside.

Option 2 removes a developer's time needed to figure out how to get hydra and the consent app talking to each other, and in my opinion a shared secret isn't that much of a degradation in security - many services are built that way anyways. But the benefit of removing the JWT complexity as well as reducing the network roundtrips is huge.

**Comment by arekkas** [comment]
@janekolszak the SDK will cover everything - it's based on swagger and works with swagger-codegen (the generated Go SDK is used in Hydra itself for the CLI and unit tested).

@ckong1991 that is really valuable feedback! What is great about option one (REST) is that we can autogenerate SDKs so you don't have to figure out anything at all, you're just calling the Hydra SDK. With autogeneration I mean using `swagger-codegen` to generate SDKs for *any* language. That way, all you have to do is:

```
hydra = NewHydraSDK("client-id", "client-secret", "http://endpoint/)

consentRequest, err := hydra.GetOAuth2ConsentRequest(r.URL.Query().Get("consent"))

// log user in and optionally check scopes

err = hydra.AcceptOAuth2ConsentRequest(r.URL.Query().Get("consent"), hydra.AcceptConsentRequestPayload{
	Subject:     "some-user-id",
	GrantScopes: []string{"some-scopes"},
})
http.Redirect(w, r, cr.RedirectUrl, http.StatusFound)
```

and that's it! This doesn't only work for Go but also for NodeJS, Java, Python, Ruby, ...

In case of JWK we would have to put a lot more effort in it, because we can't autogenerate the SDK.

Regarding request roundtrips - it doesn't matter. The reason is that the login process is slower anyways with OAuth2 since we have the redirect dance anyways (initiate oauth2 request, get forwarded to login, type user/password, get redirected back to oauth2 provider, get reditrected to app, exchange code for token(s)). 20-30ms latency more will not worsen the user experience in any meaningful way.

**Comment by ckong1991** [comment]
@arekkas ah I see. does this mean you're going to implement option 1?

**Comment by arekkas** [comment]
Yes! It will be published as 0.10.0 which is a preview to 1.0.0. I still need to get some feedback on the new flows and SDKs and improved docs which is why I don't want to publish this as a 1.0.0-prerelease because there might be some breaking changes as we progress.

I know that keeping BC is very important for developers, but it's equally important to improve developer experience. I hope you all understand this step :) I had some very good discussions with people outside of the GitHub community and a lot of people are telling me that it is more important to have good SDKs (which will now be possible) and make things as consumable as possible, even if it means that we'll break old things. However, I'll still keep updating 0.9.x with security patches if any arise!
